framework/src/mesh/MooseMesh.C

```diff
@@ -3523,72 +3535,62 @@ MooseMesh::setCoordSystem(const std::vector<SubdomainName> & blocks,
-               "'Mesh/block' parameter. Did you provide different parameter values for 'block' to "
-               "'Mesh' and 'Problem'?");
+  if (!_provided_coord_blocks.empty() && (_provided_coord_blocks != blocks))
+  {
+    const std::string param_name = isParamValid("coord_block") ? "coord_block" : "block";
+    mooseWarning("Supplied blocks in the 'setCoordSystem' method do not match the value of the "
+                 "'Mesh/",
+                 param_name,
+                 "' parameter. Did you provide different parameter values for 'Mesh/",
+                 param_name,
+                 "' and 'Problem/block'?. We will honor the parameter value from 'Mesh/",
+                 param_name,
+                 "'");
+    mooseAssert(_coord_system_set,
+                "If we are arriving here due to a bad specification in the Problem block, then we "
+                "should have already set our coordinate system subdomains from the Mesh block");
+    return;
+  }
   if (_pars.isParamSetByUser("coord_type") && getParam<MultiMooseEnum>("coord_type") != coord_sys)
-    mooseError("Supplied coordinate systems in the 'setCoordSystem' method do not match the  value "
-               "of the  'Mesh/coord_type' parameter. Did you provide different parameter values "
-               "for 'coord_type' to 'Mesh' and 'Problem'?");
-
-  const std::set<SubdomainID> & subdomains = meshSubdomains();
-  if (blocks.size() == 0)
-  {
-    // no blocks specified -> assume the whole domain
-    Moose::CoordinateSystemType coord_type = Moose::COORD_XYZ; // all is going to be XYZ by default
-    if (coord_sys.size() == 0)
-      ; // relax, do nothing
-    else if (coord_sys.size() == 1)
-      coord_type = Moose::stringToEnum<Moose::CoordinateSystemType>(
-          coord_sys[0]); // one system specified, the whole domain is going to have that system
-    else
-      mooseError("Multiple coordinate systems specified, but no blocks given.");
+    mooseError("Supplied coordinate systems in the 'setCoordSystem' method do not match the value "
+               "of the 'Mesh/coord_type' parameter. Did you provide different parameter values for "
+               "'coord_type' to 'Mesh' and 'Problem'?");
+
+  auto subdomains = meshSubdomains();
+  // It's possible that a user has called this API before the mesh is prepared and consequently we
+  // don't yet have the subdomains in meshSubdomains()
+  for (const auto & sub_name : blocks)
+  {
+    const auto sub_id = getSubdomainID(sub_name);
+    subdomains.insert(sub_id);
+  }
 
-    for (const auto & sbd : subdomains)
-      _coord_sys[sbd] = coord_type;
+  if (coord_sys.size() <= 1)
+  {
+    // We will specify the same coordinate system for all blocks
+    const auto coord_type = coord_sys.size() == 0
+                                ? Moose::COORD_XYZ
+                                : Moose::stringToEnum<Moose::CoordinateSystemType>(coord_sys[0]);
+    for (const auto sid : subdomains)
```

Comment:
@dschwen this fixes your issue regardless of the fact that `block` gets picked up from `GlobalParams`. As long as `coord_sys` is size 1 (or size 0) then we have changed to automatically set *all* mesh subdomains to that coordinate system type, regardless of the value of `block`. Before we just set it for the specifically set blocks